jenkins/Pipeline.gy

```diff
@@ -50,14 +42,54 @@ pipeline {
             }
         }
         stage('Build the Pbench Server Container') {
+            environment {
+                KEYCLOAK_CLIENT_SECRET=credentials('5cf0304d-8a00-48a4-ade5-9f59cb37ba68')
+                PB_SERVER_IMAGE_TAG="${PB_IMAGE_TAG}"
+                RPM_PATH="${WORKSPACE_TMP}/rpmbuild/RPMS/noarch/pbench-server-*.rpm"
+            }
             steps {
-                sh 'buildah login -u="${PODMAN_INTERNAL_USR}" -p="${PODMAN_INTERNAL_PSW}" ${PB_CONTAINER_REG}'
-                sh 'RPM_PATH=${WORKSPACE_TMP}/rpmbuild/RPMS/noarch/pbench-server-*.rpm bash -ex ./server/pbenchinacan/container-build-ci.sh'
-                sh 'buildah push localhost/pbench-server:${PB_SERVER_IMAGE_TAG} ${PB_CONTAINER_REG}/${PB_ORG_NAME}/pbench-server:${PB_SERVER_IMAGE_TAG}'
-                sh 'buildah push localhost/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG} ${PB_CONTAINER_REG}/${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG}'
+                sh 'buildah login -u="${PODMAN_INTERNAL_USR}" -p="${PODMAN_INTERNAL_PSW}" ${PB_INTERNAL_CONTAINER_REG}'
+                sh 'bash -ex ./server/pbenchinacan/container-build-ci.sh'
+                sh 'buildah push localhost/pbench-server:${PB_IMAGE_TAG} ${PB_INTERNAL_CONTAINER_REG}/${PB_ORG_NAME}/pbench-server:${PB_IMAGE_TAG}'
+                sh 'buildah push localhost/${PB_SERVER_IMAGE_NAME}:${PB_IMAGE_TAG} ${PB_INTERNAL_CONTAINER_REG}/${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_IMAGE_TAG}'
+            }
+        }
+        stage('Build the Pbench Agent Containers') {
+            environment {
+                // If CHANGE_ID is defined, then we are building a PR, so use
+                // the internal container registry; otherwise, we are building
```

Comment:
Why do we need to use the internal registry for PRs?

This means container images for the agent are not accessible by folks outside of Red Hat.  Why do we need to prevent that.

Why are we not just using an internal registry for the server (because we are not allowed to published publicly UBI-based images) and the external registry for the agent images?

That is to say, we only need on registry for server and one for agent.  It also reduces the complexity of this pipeline.

Reply:
> we only need on registry for server and one for agent. It also reduces the complexity of this pipeline.

Having two registries doesn't complicate the pipeline much -- we have the registry parameterized anyway, and we just set the value conditionally (which doesn't even take an extra line, IIRC).  And, as long as we have different registries for the Agent and Server, then we have two registries anyway, so I don't think it matters much.

> Why do we need to use the internal registry for PRs?

The short answer is, "I don't know".  The longer, squishier answer is that I feel like the internal registry is more under our control, using it for PR builds is less likely to confuse or disrupt our users, and doing so is less likely to cause us embarrassment of several kinds.  I envision the PR traffic as creating a lot of churn (and, possibly, resource usage), and I don't know how the folks at the external Quay.io would react to that (especially between now and when we get our management sh*t together).

That said, you bring up a valid point that external contributors would not have access to the containers that the CI would build for their PRs.  However, at this point, I believe that those contributors are unable to see _any_ of the CI results (e.g., the log!), so I'm less concerned by this than I might be.

And, given that the pipeline is fully parameterized, once we're happy with things, we can easily switch the registry for the Agent, so we don't need to do so in this PR.

> we are not allowed to published publicly UBI-based images

We need to reconfirm that -- that was my initial impression, but I believe I've since seen a statement that they are freely re-distributable...so I don't know.  Nevertheless, it's not obvious to me that we want to make our builds of the Server container freely available _anyway_, so I'm not impelled to change the Server registry at this point.

Reply:
> we are not allowed to published publicly UBI-based images

To the extent that that is true, I'm guessing that we _would_ like the ability to build the Agent in RHEL-based containers, which, if we cannot distribute them externally, would need to continue to be pushed to the internal registry and not Quay.io, which lands us back where we are now....

Reply:
Here is what I am thinking would be simpler: https://github.com/webbnh/pbench/pull/3.

Reply:
OK, we can move in that direction.  Your suggested edit looks like it has a scoping problem and a couple of stages should be combined, but we can deal with that.

I don't like burying the `credentials()` references, but it doesn't really make a difference at the point where they are actually used, it's just that finding the references (like, when we need to update them) is harder...but perhaps that's not that a big deal in the grand scheme of things.